docs/reference/style-guides/supplemental/software-architecture-implementation.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Software Architecture Implementation
+---
+
+Current as of `April 16, 2024`
+
+## Introduction
+### Motivation
+
+The layered software architecture was introduced to the team in August last year, with some literature explaining the concepts. 
+
+The primary motivation behind this document is to help better understand those concepts, by providing some practical examples, a list of do’s and don’ts and an FAQ section with common questions that came up when first implementing the concepts.
+
+### Required Reading
+
+Before continuing in this document, please make sure to read and consume the following resources. (~15min)
+
+* [Ecosystem Platform Docs - Node Style Guide - Software Architecture](https://mozilla.github.io/ecosystem-platform/reference/style-guides/node-style-guide#software-architecture)
+* [Martin Fowler - Presentation Domain Data Layering](https://martinfowler.com/bliki/PresentationDomainDataLayering.html)
+* [Martin Fowler - Repository](https://martinfowler.com/eaaCatalog/repository.html)
+
+## Guidelines
+
+This section documents the guidelines that should be followed when adopting the layered architecture.
+
+Although these aren’t hard guidelines that can't be deviated from, breaking the guidelines does carry some risks. Therefore, it is  recommended that you take a second look at your implementation and consider if breaking the rules is necessary.
+
+### List of guidelines
+
+1. Layers may not be skipped
+2. Don't go up layers
+3. Services may depend on multiple managers
+4. Managers may not depend on other managers
+5. Managers may depend on multiple repositories or clients
+6. Clients may be dependents of multiple managers
+
+### Risks of deviating from the guidelines
+
+1. Circular dependencies may occur if multiple instance on a layer become dependent on each other
+
+### Diagram of Guidelines
+
+![](../../../../static/diagrams/node-style-guide-layers.png)
+
+## Reference examples
+
+Some good examples of a Repository, Client, Manager, Service hierarchy are listed below.
+
+### End to end - libs/payments/stripe/*
+
+* Client => StripeClient ([code](https://github.com/mozilla/fxa/blob/main/libs/payments/stripe/src/lib/stripe.client.ts))
+* Manager => StripeManager ([code](https://github.com/mozilla/fxa/blob/main/libs/payments/stripe/src/lib/stripe.manager.ts))
+* Service => StripeService ([code](https://github.com/mozilla/fxa/blob/main/libs/payments/stripe/src/lib/stripe.service.ts))
```

Comment:
This is probably a bad example, because this demonstrates organization by the specific 3rd party implementation rather than organizing by feature. In organizing by feature (as shown the repository example blow), one would instead off a StripeClient, an **InvoiceManager/SubscriptionManager/CustomerManager/etc**, and then an appropriate service above based on the feature in use.